### PR TITLE
stats: allow custom histogram buckets

### PR DESCRIPTION
**Pull request by kyessenov**
Compromise to https://github.com/istio/api/pull/2539, to prefer over the bootstrap EnvoyFilter.

@zirain

### CODE REVIEW

File: annotation/annotations.yaml [hunk 1]
```diff
@@ -467,8 +476,8 @@ annotations:
 
   - name: sidecar.istio.io/extraStatTags
     featureStatus: Alpha
-    description: An additional list of tags to extract from the in-proxy Istio telemetry. each additional tag needs to be present in this list.
-    deprecated: false
+    description: An additional list of tags to extract from the in-proxy Istio Wasm telemetry. Each additional tag needs to be present in this list.
```
Comment (zirain) [inline annotation/annotations.yaml:479]: prefer to change this in a separate pr

**Review by zirain — APPROVED** [review]
this's a feature only supported by envoy, using annotation is LGTM.

File: annotation/annotations.yaml [hunk 2]
```diff
@@ -199,6 +199,15 @@ annotations:
     resources:
       - Pod
 
+  - name: sidecar.istio.io/statsHistogramBuckets
+    featureStatus: Alpha
+    description: |
+      Specifies the semicolon separated custom histogram buckets with a prefix matcher. Example: `istio:1,5,10,50,100,500,1000,5000,10000;envoy:1,5,10,25,50,100,250,500,1000,2500,5000,10000`.
```
Comment (costinm) [inline annotation/annotations.yaml:205]: - document the default 
- why do we need istio and envoy ? Can one be used for both ?

**Review by costinm — APPROVED** [review]
I think it's reasonable - we have quite a few things exposed as annotations.

One concern about pod annotations is that in Ambient they will likely cease functioning - except for Waypoints, where the annotation will go on the Gateway resource.  But that's the case for most CRDs we have with workload selector, so nothing special.

On this note - would be good to add K8S Gateway as resource that will support the annotation ( they get copied to the deployment AFAIK ).

**Comment by costinm** [comment]
Worth mentioning that customizing stats bucket for specific pods is actually best granularity for such an API - the defaults should work for most common cases, overrides are needed for pods that have unusual behavior, like long lived connections ( istiod itself ).

File: annotation/annotations.yaml (hunk 2 above)
Comment (howardjohn) [reply annotation/annotations.yaml:205]: If we do need them separate, should we just make this json? Its not that complicated but we are essentially creating a micro-language we need to parse.

if we just make them one then its a simple CSV so no problem

**Review by howardjohn — COMMENTED** [review]
Looks good overall as an experimental annotation, just one comment on the formatting (doesn't need to block, just want to make sure it is at least ACKed before moving forward)

File: annotation/annotations.yaml (hunk 2 above)
Comment (kyessenov) [reply annotation/annotations.yaml]: Updated. Prefix is not needed but usually used.

File: annotation/annotations.yaml [hunk 3]
```diff
@@ -199,6 +199,14 @@ annotations:
     resources:
       - Pod
 
+  - name: sidecar.istio.io/statsHistogramBuckets
+    featureStatus: Alpha
+    description: Specifies the custom histogram buckets with a prefix matcher to separate the Istio mesh metrics from the Envoy stats, e.g. `{"istio":{1,5,10,50,100,500,1000,5000,10000},"envoy":{1,5,10,25,50,100,250,500,1000,2500,5000,10000}}`. Default buckets are `{0.5,1,5,10,25,50,100,250,500,1000,2500,5000,10000,30000,60000,300000,600000,1800000,3600000}`.
```
Comment (howardjohn) [inline annotation/annotations.yaml:204]:
```suggestion
    description: Specifies the custom histogram buckets with a prefix matcher to separate the Istio mesh metrics from the Envoy stats, e.g. `{"istio":[1,5,10,50,100,500,1000,5000,10000],"envoy":[1,5,10,25,50,100,250,500,1000,2500,5000,10000]}`. Default buckets are `[0.5,1,5,10,25,50,100,250,500,1000,2500,5000,10000,30000,60000,300000,600000,1800000,3600000]`.
```

Current one is not valid json

Comment (howardjohn) [reply annotation/annotations.yaml:204]: can we specify what the unit is as well? milliseconds?

**Review by howardjohn — APPROVED** [review]
One small comment, otherwise LGTM. feel free to remove hold

File: annotation/annotations.yaml (hunk 3 above)
Comment (kyessenov) [reply annotation/annotations.yaml:204]: There's no unit. We have both byte count and milliseconds in use.

**Pull request merged**